src/rapi/rc_api_info.c

```diff
@@ -492,3 +492,91 @@ int rc_api_process_fetch_game_titles_server_response(rc_api_fetch_game_titles_re
 void rc_api_destroy_fetch_game_titles_response(rc_api_fetch_game_titles_response_t* response) {
   rc_buffer_destroy(&response->response.buffer);
 }
+
+/* --- Fetch Game Hashes --- */
+
+int rc_api_init_fetch_hash_library_request(rc_api_request_t* request,
+                                           const rc_api_fetch_hash_library_request_t* api_params)
+{
+  return rc_api_init_fetch_hash_library_request_hosted(request, api_params, &g_host);
+}
+
+int rc_api_init_fetch_hash_library_request_hosted(rc_api_request_t* request,
+                                                  const rc_api_fetch_hash_library_request_t* api_params,
+                                                  const rc_api_host_t* host)
+{
+  rc_api_url_builder_t builder;
+  rc_api_url_build_dorequest_url(request, host);
+
+  /* note: unauthenticated request */
+  rc_url_builder_init(&builder, &request->buffer, 48);
+  rc_url_builder_append_str_param(&builder, "r", "hashlibrary");
+  rc_url_builder_append_unum_param(&builder, "c", api_params->console_id);
+
+  request->post_data = rc_url_builder_finalize(&builder);
+  request->content_type = RC_CONTENT_TYPE_URLENCODED;
+
+  return builder.result;
+}
+
+int rc_api_process_fetch_hash_library_server_response(rc_api_fetch_hash_library_response_t* response,
+                                                      const rc_api_server_response_t* server_response)
+{
+  rc_api_hash_library_entry_t* entry;
+  rc_json_iterator_t iterator;
+  rc_json_field_t field;
+  int result;
+
+  rc_json_field_t fields[] = {
+    RC_JSON_NEW_FIELD("Success"),
+    RC_JSON_NEW_FIELD("Error"),
+    RC_JSON_NEW_FIELD("MD5List"),
+  };
+
+  memset(response, 0, sizeof(*response));
+  rc_buffer_init(&response->response.buffer);
+
+  result =
+    rc_json_parse_server_response(&response->response, server_response, fields, sizeof(fields) / sizeof(fields[0]));
+  if (result != RC_OK)
+    return result;
+
+  if (!fields[2].value_start) {
+    /* call rc_json_get_required_object to generate the error message */
+    rc_json_get_required_object(NULL, 0, &response->response, &fields[2], "MD5List");
+    return RC_MISSING_VALUE;
+  }
+
+  response->num_entries = fields[2].array_size;
+  if (response->num_entries > 0) {
+    rc_buffer_reserve(&response->response.buffer, response->num_entries * (32 + sizeof(rc_api_hash_library_entry_t)));
+
+    response->entries = (rc_api_hash_library_entry_t*)rc_buffer_alloc(
+      &response->response.buffer, response->num_entries * sizeof(rc_api_hash_library_entry_t));
+    if (!response->entries)
+      return RC_OUT_OF_MEMORY;
+
+    memset(&iterator, 0, sizeof(iterator));
+    iterator.json = fields[2].value_start;
+    iterator.end = fields[2].value_end;
+
+    entry = response->entries;
+    while (rc_json_get_next_object_field(&iterator, &field)) {
+      /* TODO: This isn't handling escape characters in the key, the RC JSON parsing functions have no method for it. */
```

Comment:
The keys should be hashes. What escape characters need to be handled?

Reply:
Can't remember for certain since I wrote this months ago, I think it was just a note because the only other places that iterate a map immediately convert the key to an integer, so there will never be any escaped characters. Whereas the key is treated as a string here, and theoretically there could be.

Practically speaking, since you control the server, probably not. Happy to drop the comment if you like.

Reply:
I'm generally not fond of using generic key maps in JSON. If this wasn't a long-standing pre-existing API, I'd have just made it an array of pairs (like the output) instead of a map.

And as long as keys aren't coming from user data, they should never need escaping. While the comment may be applicable in general, it's not applicable to the place where it's at. Please remove it.